man/man1/swilgt.1
.TH swilgt 1 "June 10, 2010" "Logtalk 2.44.1" "Logtalk Documentation"

.SH NAME
swilgt \-- Logtalk integration script for SWI-Prolog

.SH SYNOPSIS
.B swilgt
[
.BI \-g\ goal
]

.SH DESCRIPTION
\f3swilgt\f1 is a Logtalk integration script for SWI-Prolog. It automatically loads the SWI-Prolog config file, the Logtalk compiler and runtime, the Logtalk default table of library names and paths, and any existing settings file found on the startup directory or in the Logtalk user directory.

.SH OPTIONS
.TP
.BI \-g\ goal
Initialization goal, written between double quotes, to be executed after Logtalk startup.

.SH REQUIREMENTS
SWI-Prolog 5.8.0 or later is required. This integration script looks first for a "swipl" SWI-Prolog executable and, if not found, tries "pl".

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKHOME/integration/logtalk_swi.pl
Prolog file used for loading the necessary files for starting Logtalk using SWI-Prolog as the back-end compiler.
.TP
.BI $LOGTALKHOME/configs/swi.pl
The SWI-Prolog config file.
.TP
.BI $LOGTALKHOME/compiler/logtalk.pl
The Logtalk compiler and runtime file.
.TP
.BI $LOGTALKUSER/libpaths/libpaths.pl
The default table of library names and paths.
.TP
.BI $LOGTALKHOME/configs/swihooks.pl
Hooks file for supporting loading Logtalk entities using SWI-Prolog load_files/2 and consult/1 predicates and to support edit/1 and make/0.
.TP
.BI $LOGTALKHOME/configs/xpcehooks.pl
Hooks file for supporting Logtalk message sending goals as XPCE call-back goals.
.TP
.BI ./settings.lgt
Logtalk settings file (for overriding default settings in the config files). May not exist.
.TP
.BI $LOGTALKUSER/settings.lgt
Logtalk settings file (for overriding default settings in the config files). May not exist.

.SH "SEE ALSO"
The Logtalk website at http://logtalk.org/
.PP
The Logtalk User and Reference Manuals at $LOGTALKHOME/manuals/index.html
.PP
The SWI-Prolog website at http://www.swi-prolog.org/
.PP
logtalk_user_setup(1),\ logtalk_version_select(1),\ logtalk_backend_select(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2012 Paulo Moura.


RELEASE_NOTES
=================================================================
Logtalk - Object oriented extension to Prolog
Release 2.28.0

Copyright (c) 1998-2006 Paulo Moura.  All Rights Reserved.
=================================================================


RELEASE NOTES
=============


2.28.0 - June ??, 2006

	Added support for multi-threading programming on selected Prolog compilers.
	Added a new object directive, threaded/0, and four new built-in predicates,
	threaded_call/1-2 and threaded_exit/1-2. Added a new set of exemples, 
	"threads".

	Updated the config file for the forthcoming SICStus Prolog 4.0 version 
	to work in the current beta version. Modified the "make_sicstus.lgt.js" 
	script to work with both SICStus Prolog 3.12 and 4.0 versions.

	Updated the config file for YAP to workaround a problem on the Windows 
	version where testing for directory existence fails if the path ends with 
	a slash (or a double backslash); this problem prevents the use of library 
	notation to load source files.

	Dropped support for versions of XSB older than 3.0.

	Extended "parametric" and "metainterpreters" examples.

	Update all the JScript scripts to check if a compatible version of WSH is 
	installed and to workaround some problems with spaces in file paths.

	Added Inno Setup GUI Windows installer script.


2.27.1 - March 27, 2006

	Allow calls to the built-in method parameter/2 with the first argument 
	not instantiated at compile time. Detect and report as a compilation 
	error parameter/2 indexes that are out of range.

	Optimized generation of predicate definition and declaration linking 
	clauses, resulting in small lookup performance improvements and in space 
	savings for the Prolog code generated when compiling Logtalk entities.

	Many minor code and documentation improvements to the Logtalk compiler.

	Added an object version of the tokenizer described in the Michael 
	Covington's paper "Tokenization using DCG Rules" to the "dcgs" example.

	Improved integration code for Qu-Prolog and Logtalk ("configs/qphook.ql").

	Improved library hierarchy methods that return lists of objects in order 
	to avoid duplicated elements (library objects "proto_hierarchy.lgt" and 
	"class_hierarchy.lgt"). Added new methods extension/1 and extensions/1 to 
	the library object "proto_hierarchy.lgt".

	Documented the concept of "parametric object proxy". Added a new example, 
	"proxies", illustrating the use of parametric object proxies.

	Added support for code completion and for listing entity names on the 
	symbol pop-up menu to the MacOS X TextMate text editor.

	Updated the "cplgtdirs.*" scripts to also create an alias/shortcut to the 
	"wenv" directory.

	Renamed the alternative compilation and documentation directory names on 
	all config files to be compatible across operating-systems. Removed unused
	predicate '$lgt_reverse'/2 from all config files.

	Updated HTML manuals "print.css" CSS file for compatibility with the 
	latest version of CSSToXSLFO.


2.27.0 - February 9, 2006

	Improved performance for local calls to the built-in methods phrase/2-3.
	Allow the built-in methods phrase/2-3 to accept both partial and proper 
	lists of terminals.

	Improved grammar rule translator. Report calls to undefined non-terminals 
	when translating grammar rules.

	Added support for declaring grammar rule non-terminal aliases using the 
	alias/3 directive.

	Added a new predicate property, non_terminal/1, for predicates resulting 
	from the compilation of grammar rule non-terminals.

	Improved support for the built-in method expand_term/2 in order to allow 
	bypassing of the default Logtalk grammar rule translator by defining 
	clauses for the term_expansion/2 predicate. Added a "term_expansionp" 
	protocol to the Logtalk library. Added a new example, "expansion", 
	illustrating the use of the term_expansion/2 predicate.

	Added a new compiler flag, hook/1, allowing the specification of a 
	compiler hook that is called for which term read from a source file.
	Added a simple example, "hooks", of using the Logtalk compiler hook 
	to expand author abbreviations in info/1 directives into full names 
	and email addresses.

	Added support for XSB 3.x to the runtime error handler. Updated the XSB 
	3.x config file with declarations for multi-threading meta-predicates.
	
	Removed a few choice-points when checking validity of entity directives.

	Added two new objects, list(Type) and set(Type), to the standard library 
	supporting the validation of lists and ordered sets whose elements are 
	restricted to a single type.

	Added a new DCG example for solving enigmas encoded using a cellphone 
	keypad.

	Added a missing library dependency to the "puzzles" example SCRIPT file.

	Removed the experimental "systemp.lgt" protocol from the list of files 
	loaded by the "library/all_loader.lgt" loader utility files to avoid 
	compilation errors on some Prolog systems.

	Corrected a bug that prevented dynamic creation of categories using the 
	built-in predicate create_category/4.

	Corrected a bug in the reporting of singleton variables, which failed to 
	write an accurate message for facts and grammar rules.

	Corrected a bug in passing the correct calling context ("this") when 
	processing meta-calls in objects.

	Corrected a bug in scope checking with local calls to reflection and 
	database methods.

	Corrected a bug in checking the validity of the arguments of the op/3 
	directive appearing inside entities.

	Added predicates for testing if a term is a partial list or a proper list 
	to all config files.

	Added a definition for missing open/4 ISO Prolog predicate to the config 
	file of Bin Prolog.

	Added a workaround for a B-Prolog bug to this compiler config file.


2.26.2 - December 20, 2005

	Improved error-checking for the Logtalk compiler and for the grammar rule 
	translator.

	Small performance improvements for message sending and for the built-in 
	database methods.


integration/swilgt.sh
#!/bin/bash

## ================================================================
## Logtalk - Open source object-oriented logic programming language
## Release 2.44.1
## 
## Copyright (c) 1998-2012 Paulo Moura.        All Rights Reserved.
## Logtalk is free software.  You can redistribute it and/or modify
## it under the terms of the "Artistic License 2.0" as published by 
## The Perl Foundation. Consult the "LICENSE.txt" file for details.
## ================================================================

## As silly as it seems, there is no reliable solution to put the following
## checks in their own file that would be source'd within each individual  
## integration script!!!

if ! [ "$LOGTALKHOME" ]; then
	echo "The environment variable LOGTALKHOME should be defined first, pointing"
	echo "to your Logtalk installation directory!"
	echo "Trying the default locations for the Logtalk installation..."
	if [ -d "/usr/local/share/logtalk" ]; then
		LOGTALKHOME=/usr/local/share/logtalk
		echo "... using Logtalk installation found at /usr/local/share/logtalk"
	elif [ -d "/usr/share/logtalk" ]; then
		LOGTALKHOME=/usr/share/logtalk
		echo "... using Logtalk installation found at /usr/share/logtalk"
	elif [ -d "/opt/local/share/logtalk" ]; then
		LOGTALKHOME=/opt/local/share/logtalk
		echo "... using Logtalk installation found at /opt/local/share/logtalk"
	elif [ -d "/opt/share/logtalk" ]; then
		LOGTALKHOME=/opt/share/logtalk
		echo "... using Logtalk installation found at /opt/share/logtalk"
	elif [ -d "$HOME/share/logtalk" ]; then
		LOGTALKHOME="$HOME/share/logtalk"
		echo "... using Logtalk installation found at $HOME/share/logtalk"
	else
		echo "... unable to locate Logtalk installation directory!"
		echo
		exit 1
	fi
	echo
	export LOGTALKHOME=$LOGTALKHOME
elif ! [ -d "$LOGTALKHOME" ]; then
	echo "The environment variable LOGTALKHOME points to a non-existing directory!"
	echo "Its current value is: $LOGTALKHOME"
	echo "The variable must be set to your Logtalk installation directory!"
	echo
	exit 1
fi

if ! [ "$LOGTALKUSER" ]; then
	echo "The environment variable LOGTALKUSER should be defined first, pointing"
	echo "to your Logtalk user directory!"
	echo "Trying the default location for the Logtalk user directory..."
	echo
	export LOGTALKUSER=$HOME/logtalk
fi

if [ -d "$LOGTALKUSER" ]; then
	if ! [ -f "$LOGTALKUSER/VERSION.txt" ]; then
		echo "Cannot find version information in the Logtalk user directory at $LOGTALKUSER!"
		echo "Creating an up-to-date Logtalk user directory..."
		logtalk_user_setup
	else
		current=`cat $LOGTALKUSER/VERSION.txt | sed 's/\.//g'`
		if [ $current -lt 2440 ]; then
			echo "Logtalk user directory at $LOGTALKUSER is outdated!"
			echo "Creating an up-to-date Logtalk user directory..."
			logtalk_user_setup
		fi
	fi
else
	echo "Cannot find \$LOGTALKUSER directory! Creating a new Logtalk user directory"
	echo "by running the \"logtalk_user_setup\" shell script:"
	logtalk_user_setup
fi
echo

export LOGTALK_STARTUP_DIRECTORY=`pwd`

if swipl -t halt 2>&1 | grep "SWI-Prolog"; then
	exec swipl -s "$LOGTALKHOME/integration/logtalk_swi.pl" "$@"
elif pl -t halt 2>&1 | grep "SWI-Prolog"; then
	exec pl -s "$LOGTALKHOME/integration/logtalk_swi.pl" "$@"
fi
